go/consensus/tendermint/apps/roothash/roothash.go

```diff
@@ -383,13 +383,20 @@ func (app *rootHashApplication) onNewRuntime(ctx *abci.Context, runtime *registr
 	}
 
 	// Create genesis block.
-	genesisBlock := genesis.Blocks[runtime.ID]
-	if genesisBlock == nil {
-		now := ctx.Now().Unix()
-		genesisBlock = block.NewGenesisBlock(runtime.ID, uint64(now))
-		if !runtime.Genesis.StateRoot.IsEmpty() {
-			genesisBlock.Header.StateRoot = runtime.Genesis.StateRoot
+	now := ctx.Now().Unix()
+	genesisBlock := block.NewGenesisBlock(runtime.ID, uint64(now))
+	// Fill the Header fields with Genesis runtime states, if this was called during InitChain().
+	if ctx.IsInitChain() {
```

Comment:
So this is not entirely correct as it can happen that this is in InitChain but there is no `genesis.RuntimesStates` defined. In that case the genesis state supplied in the runtime descriptor should still be used (if any).